api/submit_result.go

```diff
@@ -42,24 +45,30 @@ func (s *resultSubmitter) processExecution(executionID string, outputKey string,
 
 	exec.Service, err = service.FromService(exec.Service, service.ContainerOption(s.api.container))
 	if err != nil {
-		exec.Failed(err)
-		return exec, err
+		return exec, s.failedAndSave(exec, err)
 	}
 
 	var outputDataMap map[string]interface{}
-	if err = json.Unmarshal(outputData, &outputDataMap); err != nil {
-		err = fmt.Errorf("invalid output data error: %s", err)
-		exec.Failed(err)
-	} else {
-		if err = exec.Complete(outputKey, outputDataMap); err != nil {
-			exec.Failed(err)
-		}
+	if err := json.Unmarshal(outputData, &outputDataMap); err != nil {
+		return exec, s.failedAndSave(exec, fmt.Errorf("invalid output data error: %s", err))
```

Comment:
notes:

* error here sent to workflow but same error not sent to return value of gRPC handler. gRPC handler should get the same error but it's overwritten by the error returned from _ failedAndSave()_.

Reply:
I changed the logic a little bit with the introduction of `func saveExecution`.
In `func saveExecution`, error on `exec.Failed` and `execDB.Save` are prioritize over possible error from parameters.